Elim packed

agrover opened the pull request.

Commit:
scoutfs: fix hash compiler warnings

Signed-off-by: Andy Grover <agrover@versity.com>

Commit:
scoutfs: align items in item cache to 8 bytes

This will ensure structs, which are internally 8 byte aligned, will remain
so when in the item cache.

16 bytes alignment doesn't seem like it's needed so just do 8.

Signed-off-by: Andy Grover <agrover@versity.com>

zabbo commented on src/hash.h:
```diff
@@ -26,11 +26,11 @@ static inline u32 fnv1a32(const void *data, unsigned int len)
 static inline u64 fnv1a64(const void *data, unsigned int len)
 {
-	u64 hash = 0xcbf29ce484222325;
+	u64 hash = 0xcbf29ce484222325UL;
```
a) nice catch, how did I miss that?   b) these needs to be ULL.

Commit:
scoutfs: remove __packed from all struct definitions

Instead, explicitly add padding field, and adjust member ordering to
eliminate compiler-added padding between members, and at the end of the
struct (if possible: some structs end in a u8[0] array.)

This should prevent unaligned accesses. Not a big deal on x86_64, but
other archs like aarch64 really want this.

Signed-off-by: Andy Grover <agrover@versity.com>

zabbo committed:
scoutfs: bring back sort_priv

Bring back sort_priv, we have need for sorting with a caller argument.

Signed-off-by: Zach Brown <zab@versity.com>

zabbo committed:
scoutfs: remove btree value owner footer offset

We were using a trailing owner offset to iterate over btree item values
from the back of the block towards the front.  We did this to reclaim
fragmented free space in a block to satisfy an allocation instead of
having to split the block, which is expensive mostly because it has to
allocate and free metadata blocks.

In the before times, we used to compact items by sorting items by their
offset, moving them, and then sorting them by their keys again.  The
sorting by keys was expensive so we added these owner offsets to be able
to compact without sorting.

But the complexity of maintaining the owner metadata is not worth it.
We can avoid the expensive sorting by keys by allocating a temporary
array of item offsets and sorting only it by the value offset.  That's
nice and quick, it was the key comparisons that were expensive.  Then we
can remove the owner offset entirely, as well as the block header final
free region that compaction needed.

And we also don't compact as often in the modern era because we do the
bulk of our work in the item cache instead of in the btree, and we've
changed the split/merge/compaction heuristics to avoid constantly
splitting/merging/comapcting and an item population happens to hover
right around a shared threshold.

Signed-off-by: Zach Brown <zab@versity.com>

Commit:
scoutfs: ensure btree values end on 8-byte-alignment boundary

Round val_len up to BTREE_VALUE_ALIGN (8), to keep mid_free_len aligned.

Signed-off-by: Andy Grover <agrover@versity.com>

zabbo reviewed (changes requested):
We should also add a build failure to -kmod if someone adds __packed to format.h.  We bread the build in src/Makefile if the fixed-width types make it into format.h without the __ prefix (which don't work in userspace, who knows, it was once for user/kernel alignment thing for some arch that has 32user/64kernel.. maybe sparc?).

Then the last thing I'm worried about is that we're adding all these _pad fields but we're not trying to zero them before writing them.  A lot are OK, but a lot definitely aren't.  Can we do at least a quick walk of the types that had padding added and try to zero in the write path?  I don't think getting it perfect in this pass is even possible, but we should try to address the simplest ones.  The key is terrifying.  But things like the avl (_insert and _relocate) and the inode (store_inode) are easy.  (Not having to think about any of this at all was one of the motivating efficiencies of __packed :/.))

zabbo commented on src/net.c:
```diff
@@ -330,7 +330,7 @@ static int submit_send(struct super_block *sb,
 		return -EINVAL;
 
-	msend = kmalloc(offsetof(struct message_send,
+	msend = kzalloc(offsetof(struct message_send,
```
We don't need to zero the whole allocation. Almost all of it is immediately overwritten by data from the sending caller or the receiving socket. Just set the net_header padding fields on the send side?

zabbo commented on src/btree.c:
```diff
@@ -788,6 +790,7 @@ static __le16 init_mid_free_len(int level)
 	free = SCOUTFS_BLOCK_LG_SIZE - sizeof(struct scoutfs_btree_block);
 	if (level == 0)
 		free -= SCOUTFS_BTREE_LEAF_ITEM_HASH_BYTES;
+	free = round_down(free, BTREE_VALUE_ALIGN);
```
Let's not round down the calculated hash size, creating unused block bytes.  Let's tweak the _HASH_BYTES define to create sizes that are a multiple of 8 then add a BUILD_BUG_ON in the btree that the hash_bytes is aligned to BTREE_VALUE_ALIGN.

zabbo commented on src/key.h:
```diff
@@ -78,6 +78,7 @@ static inline void scoutfs_key_set_zeros(struct scoutfs_key *key)
 	key->_sk_second = 0;
 	key->_sk_third = 0;
 	key->_sk_fourth = 0;
+	memset(key->__pad, 0, ARRAY_SIZE(key->__pad));
```
That'd be wrong if someone didn't use u8 __pad types, just use sizeof() instead of ARRAY_SIZE().

Commit:
scoutfs: Remove struct scoutfs_betimespec

Unused.

Signed-off-by: Andy Grover <agrover@versity.com>

Commit:
scoutfs: Ensure padding in structs remains zeroed

Audit code for structs allocated on stack without initialization, or
using kmalloc() instead of kzalloc().

- avl.c: zero padding in avl_node on insert.
- btree.c: Verify item padding is zero, or WARN_ONCE.
- inode.c: scoutfs_inode contains scoutfs_timespecs, which have padding.
- net.c: zero pad in net header.
- net.h: scoutfs_net_addr has padding, zero it in scoutfs_addr_from_sin().
- xattr.c: scoutfs_xattr has padding, zero it.
- forest.c: item_root in forest_next_hint() appears to either be
    assigned-to or unused, so no need to zero it.
- key.h: Ensure padding is zeroed in scoutfs_key_set_{zeros,ones}

Signed-off-by: Andy Grover <agrover@versity.com>

agrover commented:
Ok, a couple more commits that try to zero all padding when needed. My TODO list for this PR is currently empty.

zabbo commented on src/format.h:
```diff
@@ -259,14 +259,18 @@ struct scoutfs_btree_block {
 	/* leaf blocks have a fixed size item offset hash table at the end */
 };
 
+#define BTREE_VALUE_ALIGN 8
```
If we're adding this to format.h it should have a SCOUTFS_ prefix.  It's effectively a public API, it gets used in userspace too.

zabbo commented on src/btree.c:
```diff
@@ -1006,7 +1007,7 @@ static void verify_btree_block(struct super_block *sb,
 	int end_off;
 	int tot = 0;
-	int i = 0;
+	int i = 0, j = 0;
```
One definition per line, please.  (I added it to the nascent coding style text file!)

zabbo commented on src/btree.c:
```diff
@@ -1045,6 +1046,10 @@ static void verify_btree_block(struct super_block *sb,
 			goto out;
 		}
 
+		for (j = 0; i < sizeof(item->__pad); i++) {
```
for (j;i;i), oopsie doodle.

zabbo commented on src/xattr.c:
```diff
@@ -425,7 +425,7 @@ ssize_t scoutfs_getxattr(struct dentry *dentry, const char *name, void *buffer,
 	/* only need enough for caller's name and value sizes */
 	bytes = sizeof(struct scoutfs_xattr) + name_len + size;
-	xat = kmalloc(bytes, GFP_NOFS);
+	xat = kzalloc(bytes, GFP_NOFS);
```
This xat buffer in getxattr (and in list and drop) is overwritten by item data.  It doesn't need to be zeroed.  These allocations include the full item so these can be 64k.

zabbo commented on src/xattr.c:
```diff
@@ -527,7 +527,7 @@ static int scoutfs_xattr_set(struct dentry *dentry, const char *name,
 
 	bytes = sizeof(struct scoutfs_xattr) + name_len + size;
-	xat = kmalloc(bytes, GFP_NOFS);
+	xat = kzalloc(bytes, GFP_NOFS);
```
Again this can be huge.  No need to zero the whole buffer when most of it is overwritten.  Just initialize the padding in the if (value) block which prepares the xat header to be stored in items.

zabbo commented:
The fixes to squash seem fine, and just a few tweaks in the padding zeroing and it looks like we're good.

zabbo commented:
Also!  I checked for BUILD_BUG_ON()s that would be redundant if we maintain natural alignment, and didn't find any.

agrover commented:
ok should be all rebased and squashed, ready to go if you want to take another (final? fingers crossed) look.
